common/common.h

```diff
@@ -556,6 +570,215 @@ struct common_control_vector_load_info {
 // On error, returns {-1, empty}
 common_control_vector_data common_control_vector_load(const std::vector<common_control_vector_load_info> & load_infos);
 
+//
+// Antiprompt utils
+//
+
+class llama_antiprompts {
+  public:
+
+    struct llama_antiprompt {
+        std::string value;
+        bool is_grammar_trigger;
+    };
+
+    std::vector<std::string> stop_words;
+    std::vector<std::string> grammar_trigger_words;
+
+private:
+    // The Aho–Corasick algorithm allows efficient string matching with multiple patterns.
+    // See https://en.wikipedia.org/wiki/Aho%E2%80%93Corasick_algorithm
+    struct TrieNode {
```

Comment:
> ... handle multiple stop words efficiently - with grammar trigger words we may have many

Can you clarify why is that and how many stop words we could have in typical use cases?

Reply:
With functionary-v3.2 ([template](https://github.com/ochafik/llama.cpp/blob/e4d5449638b3c54957619d5dcc3a13f8a0b4324c/tests/chat/templates/meetkai-functionary-medium-v3.2.jinja#L276), [example prompt](https://github.com/ochafik/llama.cpp/blob/e4d5449638b3c54957619d5dcc3a13f8a0b4324c/tests/chat/goldens/meetkai-functionary-medium-v3.2-tool_use.txt#L62), I had to make both `tool_name\n` and `>>>tool_name\n` a trigger **for each tool** (the latter should be made to only match at the very beginning - still todo; the >>>fn occurs if the response started with a `>>>all` normal content, and/or if there's parallel tool calls). I suppose that could quickly mean a few dozen triggers.

But in less extreme / most cases, there should be only a handful of them (cf. `grammar_trigger_words.push_back` in [tool-call.cpp](https://github.com/ochafik/llama.cpp/blob/e4d5449638b3c54957619d5dcc3a13f8a0b4324c/common/tool-call.cpp#L347)).

Where the algorithm above might become more useful though is for streaming, as I'm thinking to pipe tokens into it / let it update its state machine / simplify process_token in the process maybe.

Another design option would have been to add a reluctant, non-backtracking `.*?` prefix to grammars but I couldn't find a nice way to fit it (and I dislike that it wouldn't quite work like a regexp). I also considered using the grammar to both constrain *and parse* tool calls (e.g. using naming conventions on grammar rules), but stuck to something simpler for now.

Reply:
I see. I don't have a good sense of the computation complexity for finding the trigger words in typical use cases. If we can show that the algorithm improves the performance in a measurable way, then it's ok. If not, we might want to fallback to some simpler brute-force approach.